reposerver/cache/cache.go

```diff
@@ -172,21 +187,108 @@ func (c *Cache) SetGitReferences(repo string, references []*plumbing.Reference)
+		// Skip empty data
+		if cacheItem[i][0] != "" || cacheItem[i][1] != "" {
+			res = append(res, plumbing.NewReferenceFromStrings(cacheItem[i][0], cacheItem[i][1]))
+		}
+	}
+	return &res
 }
 
-// GetGitReferences retrieves resolved Git repository references from cache
-func (c *Cache) GetGitReferences(repo string, references *[]*plumbing.Reference) error {
+// TryLockGitRefCache attempts to lock the key for the Git repository references if the key doesn't exist
+func (c *Cache) TryLockGitRefCache(repo string, lockId string) error {
+	// This try set with DisableOverwrite is important for making sure that only one process is able to claim ownership
+	// A normal get + set, or just set would cause ownership to go to whoever the last writer was, and during race conditions
+	// leads to duplicate requests
+	err := c.cache.SetItem(gitRefsKey(repo), [][2]string{{cacheutil.CacheLockedValue, lockId}}, &cacheutil.CacheActionOpts{
+		Expiration:       c.revisionCacheLockTimeout,
+		DisableOverwrite: true})
+	return err
+}
+
+func (c *Cache) GetGitReferences(repo string, lockId string) (lockOwner string, references *[]*plumbing.Reference, err error) {
 	var input [][2]string
-	if err := c.cache.GetItem(gitRefsKey(repo), &input); err != nil {
-		return err
+	err = c.cache.GetItem(gitRefsKey(repo), &input)
+	if err == ErrCacheMiss {
+		// Expected
+		return "", nil, nil
+	} else if err == nil && len(input) > 0 && len(input[0]) > 0 {
+		if input[0][0] != cacheutil.CacheLockedValue {
+			// Valid value in cache, convert to plumbing.Reference and return
+			return "", GitRefCacheItemToReferences(input), nil
+		} else {
+			// The key lock is being held
+			return input[0][1], nil, nil
+		}
 	}
-	var res []*plumbing.Reference
-	for i := range input {
-		res = append(res, plumbing.NewReferenceFromStrings(input[i][0], input[i][1]))
+	return "", nil, err
+}
+
+// GetOrLockGitReferences retrieves the git references if they exist, otherwise creates a lock and returns so the caller can populate the cache
+func (c *Cache) GetOrLockGitReferences(repo string, references *[]*plumbing.Reference) (updateCache bool, lockId string, err error) {
+	myLockUUID, err := uuid.NewRandom()
+	if err != nil {
+		log.Debug("Error generating git references cache lock id: ", err)
+		return false, "", err
 	}
-	*references = res
-	return nil
+	// We need to be able to identify that our lock was the successful one, otherwise we'll still have duplicate requests
+	myLockId := myLockUUID.String()
+	// Value matches the ttl on the lock in TryLockGitRefCache
+	waitUntil := time.Now().Add(c.revisionCacheLockTimeout)
+	// Wait only the maximum amount of time configured for the lock
+	for time.Now().Before(waitUntil) {
+		// Attempt to retrieve the key from local cache only
+		if _, cacheReferences, err := c.GetGitReferences(repo, myLockId); err != nil || cacheReferences != nil {
+			if cacheReferences != nil {
+				*references = *cacheReferences
+			}
+			return false, myLockId, err
+		}
```

Comment:
This should be called at the beginning of the function outside the for-loop in case the item already exist. 
```
func GetOrLockGitReferences() {
   GetGitReferences()
   return if cacheReferences != nil

   lock = new lock()
   for timeout {
     Trylock()
     GetGitReferences()
     return if cacheReferences != nil
     return if lockOwner == lock
  }
}
```

Reply:
I do have it purposefully inside the loop, because we are working against all controllers racing to get and set that data. The loop works like this
- Try and retrieve a valid item first (on first call even inside the loop there is no delay having it here)
- Attempt to lock the cache on the failure
- Check if the local won the lock
- Wait 1 second
- Repeat

This is a bit redundant with the two calls but the purpose is to prevent unnecessary lock set attempts. In the case of moving it out of the loop during the loop we would always would try to lock first and then do the get second. Having the redundancy is even more useful if we do end up moving to the two level cache because GET on local cache is much faster than an attempted SET on the remote

Reply:
The goal here is not to change what the function is doing, but to reduce code duplication and improve readability. Maybe my pseudo-code was more confusing than intended 😅. What do you thing of this refactor?

I think it is doing the same thing, but I personally find it more readable.

```go
func (c *Cache) GetOrLockGitReferences(repo string, references *[]*plumbing.Reference) (bool, string, error) {
	myLockUUID, err := uuid.NewRandom()
	if err != nil {
		log.Debug("Error generating git references cache lock id: ", err)
		return false, "", err
	}
	myLockId := myLockUUID.String()

	firstIteration := true
	waitUntil := time.Now().Add(c.revisionCacheLockTimeout)
	lockOwner, cacheReferences, err := c.GetGitReferences(repo)
	for time.Now().Before(waitUntil) {
		if err != nil {
			return false, "", err
		}
		if cacheReferences != nil {
			*references = *cacheReferences
			return false, "", err
		}
		if lockOwner == myLockId {
			return true, myLockId, nil
		}

		// Could not get key locally attempt to get the lock
		err = c.TryLockGitRefCache(repo, myLockId)
		if err != nil {
			// Log but ignore this error since we'll want to retry, failing to obtain the lock should not throw an error
			log.Errorf("Error attempting to acquire git references cache lock: %v", err)
		}

		lockOwner, cacheReferences, err = c.GetGitReferences(repo)
		
		if !firstIteration {
			// Wait for lock, valid value, or timeout
			time.Sleep(1 * time.Second)
		}
		firstIteration = false
	}
	// Timeout waiting for lock
	log.Debug("Repository cache was unable to acquire lock or valid data within timeout")
	return true, myLockId, nil
}
```

Reply:
@agaudreault even with this refactor the same issue persists, the GET immediately before and after the lock each time is to prevent attempting a SET on stale information. 

With your changes we would wait 1 second, attempt a SET and then run a GET. Assuming we make it past the first iteration the most likely place for another process to update the cache will be during that wait which is why the GET is always called before the SET.

To simplify it in this way the alternative would be to wait the 1 second to know if we obtained the lock by putting the GET before the SET 

It's clear that the code is too convoluted so I'll take a look at simplifying

Reply:
Restructured to simplify the loop, moved the secondary get to the lock to make that use case more obvious